src/java/main/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -522,6 +566,15 @@ protected boolean removeWatches(Map<String, Set<Watcher>> pathVsWatcher,
 
             return result;
         }
+
+        private void addPersistentWatches(String clientPath, Set<Watcher> result) {
+            synchronized (persistentWatches) {
+                PathIterator pathIterator = new PathIterator(clientPath);
+                while (pathIterator.hasNext()) {
```

Comment:
If you change ``PathIterator`` to implement both ``Iterable`` and ``Iterator`` (see below) then this line can be:

```
for (String path: pathIterator) {
    addTo(persistentWatches.get(path), result);
}
```